File: problems/993-cousins-in-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        x_parent, y_parent = None, None
        x_depth, y_depth = None, None

        def dfs(node, parent, depth):
            nonlocal x_parent, y_parent, x_depth, y_depth

            if not node:
                return

            if node.val == x:
                x_parent = parent
                x_depth = depth
            elif node.val == y:
                y_parent = parent
                y_depth = depth

            dfs(node.left, node, depth + 1)
            dfs(node.right, node, depth + 1)

        dfs(root, None, 0)

        return x_depth == y_depth and x_parent != y_parent

File: problems/993-cousins-in-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        int xDepth = -1;
        int yDepth = -1;
        TreeNode xParent = null;
        TreeNode yParent = null;

        java.util.Queue<TreeNode> queue = new java.util.LinkedList<>();
        java.util.Queue<Integer> depthQueue = new java.util.LinkedList<>();

        queue.offer(root);
        depthQueue.offer(0);

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            int depth = depthQueue.poll();

            if (node.left != null) {
                if (node.left.val == x) {
                    xDepth = depth + 1;
                    xParent = node;
                }
                if (node.left.val == y) {
                    yDepth = depth + 1;
                    yParent = node;
                }
                queue.offer(node.left);
                depthQueue.offer(depth + 1);
            }

            if (node.right != null) {
                if (node.right.val == x) {
                    xDepth = depth + 1;
                    xParent = node;
                }
                if (node.right.val == y) {
                    yDepth = depth + 1;
                    yParent = node;
                }
                queue.offer(node.right);
                depthQueue.offer(depth + 1);
            }

            if (xDepth != -1 && yDepth != -1) {
                break;
            }
        }

        return xDepth == yDepth && xParent != yParent;
    }
}